TestShardPool failure

#### Can be reproduced ?

Saw only once

#### Steps:

run `make test`

#### Expected behavior:

pass all tests

#### Actual behavior:

```
--- FAIL: TestShardPool (108.41s)
    store_shards_pool_test.go:61:
          Error Trace:  store_shards_pool_test.go:61
          Error:        Received unexpected error:
                        rpc timeout
          Test:         TestShardPool
    store_shards_pool_test.go:62:
          Error Trace:  store_shards_pool_test.go:62
          Error:        Not equal:
                        expected: 0x1
                        actual  : 0x0
          Test:         TestShardPool
    store_shards_pool_test.go:63:
          Error Trace:  store_shards_pool_test.go:63
          Error:        Not equal:
                        expected: []byte{0x70, 0x72, 0x6f, 0x70, 0x6f, 0x73, 0x65, 0x31}
                        actual  : []byte(nil)

                        Diff:
                        --- Expected
                        +++ Actual
                        @@ -1,4 +1,2 @@
                        -([]uint8) (len=8) {
                        - 00000000  70 72 6f 70 6f 73 65 31                           |propose1|
                        -}
                        +([]uint8) <nil>

          Test:         TestShardPool
    testutil.go:763:
          Error Trace:  testutil.go:763
                              store_shards_pool_test.go:64
          Error:
          Test:         TestShardPool
          Messages:     wait shard state changed to Running timeout
```


#### Environment:
- Version or commit-id (e.g. v0.1.0 or 8b23a93):
- Hardware parameters:
- OS type:
- Others:

#### Configuration file:


#### Additional context:
- Error message from client:
- Server log:
- Other information:

--- a/components/prophet/config/config.go
+++ b/components/prophet/config/config.go
@@ -64,6 +64,7 @@ type Config struct {
 		jobProcessors map[metapb.JobType]JobProcessor
 	} `toml:"-" json:"-"`
 
+	// TODO(fagongzi): the following test-related configurations are moved to a separate struct
 	// Only test can change them.
 	DisableStrictReconfigCheck bool `toml:"-" json:"-"`
 	// DisableResponse skip all client request

--- a/config/config.go
+++ b/config/config.go
@@ -430,6 +430,8 @@ type TestConfig struct {
 	PeerReplicaSetSnapshotJobWait time.Duration
 	// SaveDynamicallyShardInitStateWait wait before save dynamically shard init state
 	SaveDynamicallyShardInitStateWait time.Duration
+	// ShardPoolCreateWaitC waiting delay for shard creation
+	ShardPoolCreateWaitC chan struct{}
 	// Shards test config for shards
 	Shards map[uint64]*TestShardConfig
 }

--- a/raftstore/store_shards_pool.go
+++ b/raftstore/store_shards_pool.go
@@ -21,7 +21,8 @@ import (
 )
 
 var (
-	errNoIdleShard = errors.New("no idle shard")
+	errNoIdleShard   = errors.New("no idle shard")
+	batchCreateCount = 4
 )
 
 // ShardsPool is a shards pool, it will always create shards until the number of available shards reaches the
@@ -49,6 +50,7 @@ func (s *store) CreateResourcePool(pools ...metapb.ResourcePool) (ShardsPool, er
 
 // dynamicShardsPool a dynamic shard pool
 type dynamicShardsPool struct {
+	cfg     *config.Config
 	factory func(g uint64, start, end []byte, unique string, offsetInPool uint64) bhmetapb.Shard
 	job     metapb.Job
 	ctx     context.Context
@@ -66,7 +68,7 @@ type dynamicShardsPool struct {
 }
 
 func newDynamicShardsPool(cfg *config.Config) *dynamicShardsPool {
-	p := &dynamicShardsPool{pdC: make(chan struct{}, 1)}
+	p := &dynamicShardsPool{pdC: make(chan struct{}, 1), cfg: cfg}
 	p.factory = p.shardFactory
 	if cfg.Customize.CustomShardPoolShardFactory != nil {
 		p.factory = cfg.Customize.CustomShardPoolShardFactory
@@ -305,39 +307,57 @@ func (dsp *dynamicShardsPool) triggerCreateLocked() {
 }
 
 func (dsp *dynamicShardsPool) maybeCreate(store storage.JobStorage) {
-
 	dsp.mu.Lock()
-	defer dsp.mu.Unlock()
-
 	if !dsp.isStartedLocked() {
+		dsp.mu.Unlock()
 		return
 	}
 
-	old := dsp.cloneDataLocked()
+	// we don't modify directly
+	modified := dsp.cloneDataLocked()
 	var creates []metadata.Resource
-	for g, p := range dsp.mu.pools.Pools {
-		if p.Seq == 0 ||
-			(int(p.Seq-p.AllocatedOffset) < int(p.Capacity) && len(creates) < 8) {
-			p.Seq++
-			creates = append(creates, NewResourceAdapterWithShard(dsp.factory(g,
-				addPrefix(p.RangePrefix, p.Seq),
-				addPrefix(p.RangePrefix, p.Seq+1),
-				dsp.unique(g, p.Seq),
-				p.Seq)))
+	for {
+		changed := false
+		for g, p := range modified.Pools {
+			if p.Seq == 0 ||
+				(int(p.Seq-p.AllocatedOffset) < int(p.Capacity) && len(creates) < batchCreateCount) {
+				p.Seq++
+				creates = append(creates, NewResourceAdapterWithShard(dsp.factory(g,
+					addPrefix(p.RangePrefix, p.Seq),
+					addPrefix(p.RangePrefix, p.Seq+1),
+					dsp.unique(g, p.Seq),
+					p.Seq)))
+				changed = true
+			}
+		}
+
+		if !changed {
+			break
 		}
 	}
+	dsp.mu.Unlock()
 
 	if len(creates) > 0 {
+		if dsp.cfg.Test.ShardPoolCreateWaitC != nil {
+			<-dsp.cfg.Test.ShardPoolCreateWaitC
+		}
 		err := dsp.pd.AsyncAddResources(creates...)
 		if err != nil {
 			logger.Errorf("shards pool create shards failed with %+v", err)
-			dsp.mu.pools = old
 			return
 		}
 
+		// only update seq, all operation to update seq are in the same goroutine.
+		dsp.mu.Lock()
+		defer dsp.mu.Unlock()
+
+		backup := dsp.cloneDataLocked()
+		for g, p := range dsp.mu.pools.Pools {
+			p.Seq = modified.Pools[g].Seq
+		}
 		if err := dsp.saveLocked(store); err != nil {
-			logger.Errorf("save shard pool job data failed with %+v", err)
-			dsp.mu.pools = old
+			logger.Errorf("save shard pool job data failed with %+v, retry later", err)
+			dsp.mu.pools = backup
 		}
 
 		dsp.triggerCreateLocked()

--- a/raftstore/store_shards_pool_test.go
+++ b/raftstore/store_shards_pool_test.go
@@ -2,6 +2,7 @@ package raftstore
 
 import (
 	"fmt"
+	"strings"
 	"testing"
 	"time"
 
@@ -142,4 +143,25 @@ func TestShardPoolWithFactory(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, "b1", string(resp["r1"].Responses[0].Value))
 	assert.Equal(t, "b2", string(resp["r2"].Responses[0].Value))
+}
+
+func TestIssue192(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	wc := make(chan struct{})
+	c := NewSingleTestClusterStore(t, WithAppendTestClusterAdjustConfigFunc(func(i int, cfg *config.Config) {
+		cfg.Customize.CustomInitShardsFactory = func() []bhmetapb.Shard { return []bhmetapb.Shard{{Start: []byte("a"), End: []byte("b")}} }
+		cfg.Test.ShardPoolCreateWaitC = wc
+	}))
+	defer c.Stop()
+
+	c.Start()
+
+	p, err := c.stores[0].CreateResourcePool(metapb.ResourcePool{Group: 0, Capacity: 1, RangePrefix: []byte("b")})
+	assert.NoError(t, err)
+	assert.NotNil(t, p)
+
+	_, err = p.Alloc(0, []byte("purpose"))
+	assert.Error(t, err)
+	assert.False(t, strings.Contains(err.Error(), "timeout"))
 }

